criu/files-reg.c

```diff
@@ -1792,30 +1792,49 @@ static int linkat_hard(int odir, char *opath, int ndir, char *npath, uid_t uid,
 	return ret;
 }
 
-static void rm_parent_dirs(int mntns_root, char *path, int count)
+/*
+ * FIXME these function changes path string in place, if this string is used
+ * simultaneousely in multiple processes we can have a race. Do we need strdup?
```

Comment:
Yes, it's better to be safe

Reply:
Something like https://github.com/Snorch/criu/commit/98f0445f884e3b2317d9efc74a77fd0a83bf48c9 can be done, but I'm not sure that we urgently need this because we've never seen such a race and all uses of path seem local not shared. But for future we should have this in mind.

Reply:
Even if it's not racy, i would do it.

It's very surprising that a function that unlinks things actually modifies an argument. It makes the code more difficult to reason about, and thus more difficult to maintain.

Correctness is more important that saving a few cpu cycles

Reply:
> Correctness is more important that saving a few cpu cycles

Excess correctness checks are not always welcome =) One example that I can remember is --check-mounts patch (https://github.com/checkpoint-restore/criu/pull/660#issuecomment-548736668).

But anyway I tried to fix this comment. PR become a bit more complex. I'm OK to revert it to the original form in case @checkpoint-restore/maintainers are against this "strdup and staff" change.

Reply:
> Excess correctness checks are not always welcome
I agree. I don't think using defensive programming is useful in most cases.

Here, we are calling `rm_parent_dirs(mntns_root, rfi->path, level)`. I'm willing to bet that most programmers will say that It's unexpected that this call modifies the rfi->path string for two reasons 1) semantically, there's no reason that unlink()/rm_dirs() modifies its arguments 2) we are modifying the content of a struct. There's no reason to think that this string is temporary or won't be reused.

This can lead to future bugs. I stand strong on this stance.

For implementation details, it can be easier to reason about it by putting the dup logic outside of the function:

```c
int __rm_parent_dirs(..., char *path, ...) {
    ...
}

int rm_parent_dirs(..., const char *path, ...) {
    char *dup_path = strdup(path);
    if (!dup_path)
        return -1;
    int ret = __rm_parent_dirs(..., dup_path, ...);
    free(dup_path);
    return ret;
}
```

Reply:
> Here, we are calling rm_parent_dirs(mntns_root, rfi->path, level). I'm willing to bet that most programmers will say that It's unexpected that this call modifies the rfi->path string for two reasons 1) semantically, there's no reason that unlink()/rm_dirs() modifies its arguments 2) we are modifying the content of a struct. There's no reason to think that this string is temporary or won't be reused.

All modifications of rm_parent_dirs are invisible to the caller, it only temporary inserts '\0' in place of '\' and reverts everything back later.

> This can lead to future bugs. I stand strong on this stance.

That's why I initially wanted to add FIXME comment, so that it would be easier to track such bugs in case we would have them in future. And now I tend to think that fixing rm_parent_dirs argument modification can be a matter of another PR. This pr goal is to fix remount writable code, and fix error handling so that the problem with remount writable code would be visible (test mntn_ghost01 starts to fail).

> For implementation details, it can be easier to reason about it by putting the dup logic outside of the function:

The Idea looks good, but as we need to intrusively change rm_parent_dirs making it not replace '\0' back to '\' as this code becomes excess, so I don't think we need __rm_parent_dirs ...

Let's be lazy an leave strdup like I did it, and wait for some more review.

Reply:
> All modifications of rm_parent_dirs are invisible to the caller, it only temporary inserts '\0' in place of '\' and reverts everything back later.

Ohh okay, well in that case, it's fine. Nevermind.